middleware/dnstap/handler.go

```diff
@@ -0,0 +1,39 @@
+package dnstap
+
+import (
+	"golang.org/x/net/context"
+	"io"
+
+	"github.com/coredns/coredns/middleware"
+	"github.com/coredns/coredns/middleware/dnstap/msg"
+
+	tap "github.com/dnstap/golang-dnstap"
+	"github.com/miekg/dns"
+	"github.com/pkg/errors"
+)
+
+type ClientTap struct {
+	Next middleware.Handler
+	Out  io.Writer
+	Pack bool
+}
+
+func (h ClientTap) TapMessage(m *tap.Message) error {
+	frame, err := msg.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "marshal")
+	}
+	_, err = h.Out.Write(frame)
+	return err
+}
+
+func (h ClientTap) ServeDNS(ctx context.Context, w dns.ResponseWriter,
+	r *dns.Msg) (int, error) {
+
+	taprw := ResponseWriter{ResponseWriter: w, Tap: &h, Query: r, Pack: h.Pack}
+	taprw.QueryEpoch()
```

Comment:
The epoch value of the query request, without this method, the info would be lost when `ResponseWriter.Write` is called and logs both the query and response.